### ISSUE TITLE
Generic functions with jax arrays, and incoherent `issubclass` and `isinstance` for `jnp.ndarray`

### ISSUE THREAD

**Issue by Gattocrucco** [google/jax#11075]
**TLDR:** The Python type dispatch mechanism for generic functions does not work within jax tracing with `jnp.ndarray` as type. Making `issubclass(Tracer, jnp.ndarray)` come out true is still not sufficient.

I often have to work around the fact that jax arrays cannot be dealt with by numpy ufuncs because I have to write functions which accept both kind of arrays. I tried to use the Python standard [dispatch mechanism](https://docs.python.org/3.10/library/functools.html#functools.singledispatch) but failed when tracing.

```python
import functools

import jax
import jax.numpy as jnp

@functools.singledispatch
def f(_):
    print('failure')

@f.register
def _(_: jnp.ndarray):
    print('success')

f(jnp.zeros(1)) # -> success

def g(x):
    f(x)
    return x

jax.jacobian(g)(jnp.zeros(1)) # -> failure
```

This was unexpected because I used many times `isinstance(x, jnp.ndarray)` in traced code successfully. So I found out that `jax.core.Tracer` is not a subclass of `jnp.ndarray`, but the instance checking is overridden:

```python
class MyTracer(jax.core.Tracer):
    __init__ = lambda _: None
    aval = property(lambda _: jax.core.UnshapedArray(object))

print(isinstance(MyTracer(), jnp.ndarray)) # -> True
print(issubclass(jax.core.Tracer, jnp.ndarray)) # -> False
```

The code for `type(jnp.ndarray)` is:

```python
class ArrayMeta(abc.ABCMeta):
  """Metaclass for overriding ndarray isinstance checks."""

  def __instancecheck__(self, instance):
    # Allow tracer instances with avals that are instances of UnshapedArray.
    # We could instead just declare Tracer an instance of the ndarray type, but
    # there can be traced values that are not arrays. The main downside here is
    # that isinstance(x, ndarray) might return true but
    # issubclass(type(x), ndarray) might return false for an array tracer.
    try:
      return (hasattr(instance, "aval") and
              isinstance(instance.aval, core.UnshapedArray))
    except AttributeError:
      super().__instancecheck__(instance)
```

Given the comment complaining about the inconsistency, why isn't `__subclasscheck__` implemented? I guess that the difficulty is in distinguishing tracer classes that trace arrays, probably right now the same classes can trace different kinds of objects. Anyway, putting this aside, I add a `__subclasscheck__` that assumes everything is an array:

```python
ArrayMeta = type(jnp.ndarray)
def __subclasscheck__(self, subclass):
    if issubclass(subclass, jax.core.Tracer):
        return True
    else:
        return super(ArrayMeta, self).__subclasscheck__(subclass)
ArrayMeta.__subclasscheck__ = __subclasscheck__
print(issubclass(jax.core.Tracer, jnp.ndarray)) # -> True
```

However, this still does not fix the dispatch type matching mechanism:

```python
jax.jacobian(g)(jnp.zeros(1)) # -> failure
```

Of course now that I understand how this works, I can manually add the tracer type:

```python
@f.register
def _(_: jax.core.Tracer):
    print('~success')

jax.jacobian(g)(jnp.zeros(1)) # -> ~success
```

But this will suffer from the same inconsistency due to non-array tracers. So in practice I write my own wrappers using `isinstance`.

**Comment by jakevdp** [comment]
As you've realized, the issue comes down to the fact that JAX heavily uses duck-typing in its transforms (i.e. replacing true arrays with tracers in the execution of functions), and thus is not directly compatible with APIs that rely on strict type checking. We've hacked around that a bit for `isinstance`, but I don't imagine anyone believes that's a comprehensive fix to making JAX work with interfaces that rely on strong types.

If you have specific ideas of how JAX could be improved to more closely emulate strong types within its duck-typed API, I think we'd be happy to review pull requests. But if that proves difficult, I think our response would be that it's working as intended given the duck-typed design.

What do you think?

**Comment by Gattocrucco** [comment]
Ok, granted that I don't know jax internals so I may be saying something which breaks into a flaming mess, I would create dummy subclasses matching the traced types, and then expose unions concrete type+tracer type to be used in typing. Example: `JVPTracer` would have a subclass `JVPArrayTracer`. This subclass makes no additional definition. `JVPTracer` instead of an `__init__` has a `__new__` that decides the subclass to use based on the type of `primal`. Then a submodule `jax.types` provides the definition `JaxArray = Union[jnp.ndarray, JVPArrayTracer, BatchArrayTracer, ...]`. This could be automatized with a registry that creates subclasses and corresponding unions on demand for arbitrary types and which is used by the constructors of the tracers.

About the pull request: I won't do it, it's not a priority, it was immediate to write a custom wrapper with `isinstance` (thank you for making at least `isinstance` work), but I thought it was polite to make a bug report for reference.

**Comment by jakevdp** [comment]
Interesting - I think the real problem here is that Python has two meanings of the word "type": there's "type" in the sense of `isinstance`, and "type" in the sense of `mypy`-style static type checking, and in all but the most common of cases, details of how the latter are handled can differ depending on whether you're using mypy, pytype, or pyright. So to make any change like this, you'd first need to decide what kind of "type" you're interested in, secondly you'd have to think through all the consequences of any such change for the other types of types.

With that framing, I'm a bit confused about your goal here, because initially it seemed you were interested in type annotations for multiple dispatch (which are types of the `isinstance` sort). But in your proposal, you talk about `Union`, which are types of the `mypy` sort. The two can sometimes overlap, but in general their properties and requirements differ (which is why, e.g. numpy has `np.ndarray` for types of the `isinstance` sort, and `np.typing.ArrayLike` for types of the mypy sort).

**Comment by Gattocrucco** [comment]
You are right, I naively expected `functools.singledispatch` to understand `typing.Union`s as OR of classes since it supports type annotations, but it actually doesn't:

```python
import functools
import typing

@functools.singledispatch
def f(_):
    print('not dispatched')

@f.register
def _(_: typing.Union[str, int]):
    print('dispatched')

f(None)
f('')
f(0)
```

```
TypeError: Invalid annotation for '_'. typing.Union[str, int] is not a class.
```

And since in the starting example I determined that `singledispatch` does not care about `isinstance` neither `issubclass`, there really needs to be actual inheritance. Modifying the scheme above, JAX would make `JVPArrayTracer` a subclass of both `JVPTracer` and an empty class `JaxArray`, which is also superclass of `jnp.ndarray`. With multiple inheritance you can add dummy superclasses around as you wish.

**Comment by jakevdp** [comment]
I don't think it will be possible or desirable to rewrite JAX so that tracers and arrays exist in the same object hierarchy. JAX is specifically built on duck-typing, and any programming patterns that are incompatible with duck-typing are probably incompatible with JAX.

**Comment by YouJiacheng** [comment]
@Gattocrucco I tried your `__subclasscheck__ ` method(I figured out this method according to the source code of singledispatch, and I found that you have pointed it out), and it worked!

```python
from functools import singledispatch

import jax
import jax.numpy as jnp

ArrayMeta = type(jnp.ndarray)
def __subclasscheck__(self, subclass):
    if issubclass(subclass, jax.core.Tracer):
        return True
    else:
        return super(ArrayMeta, self).__subclasscheck__(subclass)
ArrayMeta.__subclasscheck__ = __subclasscheck__


@singledispatch
def f(_):
    print('failure')

@f.register
def _(x: jnp.ndarray):
    print('success')
    return x

jax.jacobian(f)(jnp.zeros(1)) # success
```

**Comment by jakevdp** [comment]
Great - would you like to put together a PR?

**Comment by YouJiacheng** [comment]
@jakevdp The main concern is "We could instead just declare Tracer an instance of the ndarray type, but there can be traced values that are not arrays."
Duck-type is okay, but "per-instance duck-type"(i.e. different instance of same class can have different duck-type) will prevent class level check.
It seems that `DynamicJaxprTracer.aval` completely depends on instance attribute.

**Comment by YouJiacheng** [comment]
If the purpose is just make `functools.singledispatch` work, we can override `obj.__class__` behavior of Tracer
Proof of concept code:
```python
from functools import singledispatch


class A:
    pass

class B:
    def __init__(self, aval_ducktype: type = None):
        self.aval_ducktype = aval_ducktype
    
    @property
    def __class__(self):
        return self.aval_ducktype or B

@singledispatch
def f(_):
    print('not dispatched')

@f.register
def _(_: A):
    print('A')

@f.register
def _(_: B):
    print('B')

@f.register
def _(_: int):
    print('int')

f(A()) # A
f(B()) # B
f(B(A)) # A
f(B(int)) # int
f(B(str)) # not dispatched
print(isinstance(B(A), A)) # True
```
This technique is also used by stdlib: https://docs.python.org/3/library/weakref.html#weakref.proxy
```python
import weakref
from functools import singledispatch

class A: pass

@singledispatch
def f(_): print('not dispatched')
@f.register
def _(_: A): print('A')

a = A()
p = weakref.proxy(a)
print(type(p)) # <class 'weakproxy'>
print(vars(object)['__class__'].__get__(p)) # <class 'weakproxy'>
print(p.__class__) # <class '__main__.A'>
print(isinstance(p, A)) # True
f(p) # A
```
reference: https://stackoverflow.com/a/53757539/16613821

**Comment by Gattocrucco** [comment]
Cool! I must have been doing something wrong while trying `__subclasscheck__`. Anyway, your solution with `__class__` is more elegant.

**Comment by Gattocrucco** [comment]
Closing because of `jax.Array` since version 0.4.0.
